### ISSUE TITLE
Debugging webpack 5 cache issue

### ISSUE THREAD

**Issue by otakustay** [webpack/webpack#10707]
## Feature request

We were testing webpack 5's cache system to boost our build performance, we found it works in local machines where build time is reduced to some seconds, however in our CI/CD machine it still causes around half of the full build time.

We run `tar -cfz build-cache.tar.gz node_modules/.cache/webpack` and upload this tar file to cloud storage, download it on next build.

Even though we confirmed `cache.version` and `cache.name` are all identical and source code is not modified, it still cause about 4-5 minutes to finish build, comparing to 10min of a fresh build.

**What is the expected behavior?**

We'd like to have a more verbose log including how cache is read, used and deprecated during build so we can find where the build speed issue lies.

**What is motivation or use case for adding/changing the behavior?**

Since webpack is quite a blackbox, we are currently blocked from further optimize our build speed, and a detailed log can help.

**How should this be implemented in your opinion?**

Yes, a more detailed log is helpful to optimize webpack build.

**Are you willing to work on this yourself?**

Sorry but I'm not familiar with this.

**Comment by vankop** [comment]
Thanks for creating issue

Did you measure non webpack stuff in CI? I mean how much time you spend to gzip/unzip and uploading/downloading cache?

**Comment by otakustay** [comment]
I have already excluded time to download, extract and compress, I collect timestamps from each line of log and found a large gap between these 2 lines:

```text
[2020-04-13 08:59:55] Cache build-production in /foo/bar with key aaa3e72f7035e5d91ff766c445ff89114fe49f6f # This is cache.version
[2020-04-13 08:59:58] (node:22726) [DEP_WEBPACK_MAIN_TEMPLATE_GET_ASSET_PATH] DeprecationWarning: MainTemplate.getAssetPath is deprecated (use Compilation.getAssetPath instead)
[2020-04-13 09:04:06] (node:22726) [DEP_WEBPACK_MAIN_TEMPLATE_GET_PUBLIC_PATH] DeprecationWarning: MainTemplate.getPublicPath is depreacted (use Compilation.getAssetPath(compilation.outputOptions.publicPath, options) instead)
```

I confirmed logged `cache.version` are always the same, I compute this version from `process.cwd`, `node_modules/.yarn-integrity` and many others.

After this big gap is some warning about serializing and deserializing big strings, which caused about 30 seconds, which is perfectly acceptable.

**Comment by otakustay** [comment]
In my local machine it is super fast between these 2 lines:

![image](https://user-images.githubusercontent.com/639549/79111482-30d16700-7daf-11ea-94de-7a36a4bc6c58.png)

I believe webpack is building the application between `getAssetPath` and `getPublicPath` are called

**Comment by vankop** [comment]
Better if you could minimize reproduction.. Also information where deprecation happen could help. 

Try to run build with `--trace-deprecation` flag, e.g.:
```
node --trace-deprecation node_modules/.bin/webpack-cli --env.development
```

**Comment by otakustay** [comment]
I will try my best, but since we have a quite complex webpack config encapsulated in another CLI tool, it is difficult.

Is there any document describes how webpack's cache file is formatted? I wonder we can get more information after unpack cache files to a human readable one.

**Comment by vankop** [comment]
> Is there any document describes how webpack's cache file is formatted?

It described in each middleware separately:
https://github.com/webpack/webpack/blob/master/lib/serialization/FileMiddleware.js#L15
https://github.com/webpack/webpack/blob/master/lib/serialization/BinaryMiddleware.js#L13
https://github.com/webpack/webpack/blob/master/lib/serialization/ObjectMiddleware.js#L23

But I think it wont be useful..



> I wonder we can get more information after unpack cache files to a human readable one.

There is a script for printing packs  https://github.com/webpack/webpack/blob/master/tooling/print-cache-file.js

**Comment by otakustay** [comment]
I created an experiment to reproduce this issue

1. Have a success build
2. Copy `node_modules/.cache` to another location
3. Delete current project directory
4. Run `git clone` again to create a fresh project
5. `yarn install` to install dependencies
6. Copy back `node_modules/.cache` folder
7. Run build again

The key step is to delete and re-clone project, after that build becomes slow, I wonder is it possible that ctime and mtime of file impact cache usage?

**Comment by vankop** [comment]
> I wonder is it possible that ctime and mtime of file impact cache usage?

Looks like yes, see https://github.com/webpack/webpack/blob/master/lib/FileSystemInfo.js#L1478

**Comment by otakustay** [comment]
Will it be possible to have a configuration to disable this behavior, or to have a customizable algorithm to identify files version so that we can use MD5 hash or anything else? In CI/CD like GitHub Actions it is common to have a fresh clone every time, sharing cache between builds can significantly speed up development

**Comment by otakustay** [comment]
A more compatible way could be allowing user to customize timestamp algorithm, in this case we can determine mtime from git log to stabilize it

**Comment by sokra** [comment]
Yep, a hybrid snapshotting mode was already proposed. Instead of only relying on mtime, we instead could check mtime first and fallback to comparing hash of file content when mtime differ.

**Comment by sokra** [comment]
regarding debugging these issues:

```
stats: {
  loggingDebug: /FileSystemInfo/
}
```

This gives good information about which file causes snapshot invalidation.

**Comment by otakustay** [comment]
Thanks @vankop @sokra 

For anyone encountering the same issue, I've written a short script to change file's mtime according to git log

```shell
# change git managed file's mtime to last committed
git ls-tree -r --name-only HEAD | while read filename; do
  touch -amt "$(git log -1 --pretty="format:%ad" --date=format:%Y%m%d%H%M.%S $filename)" "$filename"
done

# change files unders node_modules to last time yarn.lock committed
# change it to package-lock.json if npm is used
yarn_lock_mtime=$(git log -1 --pretty="format:%ad" --date=format:%Y%m%d%H%M.%S yarn.lock)
find node_modules -type f | xargs touch -mt $yarn_lock_mtime
```

**Comment by otakustay** [comment]
Another problem is that webpack's cache relies on the full path of file, however in some CI/CD system repository is cloned into a random directory on each build, causing full path changes and cache invalidates, is there any solution to get rid of full path dependencies?

**Comment by vankop** [comment]
> Another problem is that webpack's cache relies on the full path of file

There is issue and PR for that https://github.com/webpack/webpack/issues/10400 .

**Comment by vankop** [comment]
still valid

**Comment by sokra** [comment]
should be fixed by #11379
